### PR TITLE
Add Cumulative aggregation

### LINKED ISSUE #5215: Add an Cumulative aggregation, similar to Rolling
**Is your feature request related to a problem? Please describe.**

Pandas has a [`.expanding` aggregation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.expanding.html), which is basically rolling with a full lookback. I often end up supplying rolling with the length of the dimension, and this is some nice sugar for that. 

**Describe the solution you'd like**
Basically the same as pandas — a `.expanding` method that returns an `Expanding` class, which implements the same methods as a `Rolling` class.

**Describe alternatives you've considered**
Some options:
– This
– Don't add anything, the sugar isn't worth the additional API.
– Go full out and write specialized expanding algos — which will be faster since they don't have to keep track of the window. But not that much faster, likely not worth the effort.

### PR DESCRIPTION
**Pull request by max-sixty**
Closes #5215

### CODE REVIEW

File: doc/api.rst [hunk 1]
```diff
@@ -182,6 +182,7 @@ Computation
    Dataset.groupby_bins
    Dataset.rolling
    Dataset.rolling_exp
+   Dataset.cumulative
```
Comment (dcherian) [inline doc/api.rst:185]: Thoughts on using a verb like `accumulate`?

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -6895,14 +6895,90 @@ def rolling(
...
         dim = either_dict_or_kwargs(dim, window_kwargs, "rolling")
         return DataArrayRolling(self, dim, min_periods=min_periods, center=center)
 
+    def cumulative(
+        self,
+        dim: str | Iterable[Hashable],
+        min_periods: int | None = 1,
+    ) -> DataArrayRolling:
+        """
+        Rolling window object for DataArrays.
```
Comment (dcherian) [inline xarray/core/dataarray.py:6913]:
```suggestion
        Accumulating object for DataArrays.
```
?

File: xarray/core/dataset.py [hunk 3]
```diff
@@ -10270,14 +10270,60 @@ def rolling(
...
+            dim = {dim: self.sizes[dim]}
+        else:
+            missing_dims = set(dim) - set(self.dims)
+            if missing_dims:
+                raise ValueError(
+                    f"Dimensions {missing_dims} not found in data dimensions: {self.dims}"
+                )
+            dim = {d: self.sizes[d] for d in dim}
+
+        return DatasetRolling(self, dim, min_periods=min_periods or 1, center=False)
```
[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/core/dataset.py:10325]: Where do you see this padding? I agree if that's the case. But it should just be a wrapper for `numbagg.move_sum` (currently an inelegant wrapper, but that's another refactor)

File: doc/api.rst (hunk 1 above)
Comment (max-sixty) [reply doc/api.rst:185]: I see it as a bit too numpy-lingo-ish. 

I quite liked `.cumulutive` (I think it was your idea!) — it's the normal way of describing the calc...

Has anyone every said "accumulate the sum"? vs "take a cumulative sum"...

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -10369,14 +10369,60 @@ def rolling(
...
         dim = either_dict_or_kwargs(dim, window_kwargs, "rolling")
         return DatasetRolling(self, dim, min_periods=min_periods, center=center)
 
+    def cumulative(
+        self,
+        dim: str | Iterable[Hashable],
+        min_periods: int | None = 1,
+    ) -> DatasetRolling:
+        """
+        Rolling window object for Datasets
```
Comment (max-sixty) [inline xarray/core/dataset.py:10387]:
```suggestion
        Accumulating object for Datasets
```

[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/core/dataset.py]: D'oh! Fixed

**Comment by max-sixty** [comment]
OK, awaiting your review @dcherian !

File: xarray/core/dataarray.py [hunk 5]
```diff
@@ -6923,14 +6923,90 @@ def rolling(
...
         """
         from xarray.core.rolling import DataArrayRolling
 
         dim = either_dict_or_kwargs(dim, window_kwargs, "rolling")
         return DataArrayRolling(self, dim, min_periods=min_periods, center=center)
 
+    def cumulative(
+        self,
+        dim: str | Iterable[Hashable],
+        min_periods: int | None = 1,
```
Comment (Illviljan) [inline xarray/core/dataarray.py:6938]: What does None do?

File: xarray/tests/test_rolling.py [hunk 6]
```diff
@@ -809,6 +827,25 @@ def test_raise_no_warning_dask_rolling_assert_close(self, ds, name) -> None:
         expected = getattr(getattr(ds.rolling(time=4), name)().rolling(x=3), name)()
         assert_allclose(actual, expected)
 
+    @pytest.mark.parametrize("func", ["mean", "sum"])
+    @pytest.mark.parametrize("ds", (2,), indirect=True)
+    @pytest.mark.parametrize("min_periods", [None, 1, 10])
+    def test_cumulative(self, ds, func, min_periods) -> None:
+        # One dim
+        result = getattr(ds.cumulative("time", min_periods=min_periods), func)()
```
Comment (dcherian) [inline xarray/tests/test_rolling.py:835]: Can we compare to `cumsum` for the 1D case please?

Comment (max-sixty) [reply xarray/tests/test_rolling.py:835]: Yes good idea

File: xarray/core/dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (max-sixty) [reply xarray/core/dataset.py:10424]: Yeah, for cumulative, `center=False` always, so I don't think we hit that at all...

**Pull request merged**